### DISCUSSION TITLE
Why do you like full-width code mode

### DISCUSSION THREAD

**Discussion by iHiD** [exercism/v3#2169] [General]
Having full-width code for mentors dramatically reduces the amount of options we have with building a new UX. So we're trying to understand whether it's truly necessary or if there might be better ways we can do it.

For those people who have full-width enabled, could you please outline why below. Preferably with an example solution or two for us to look at. Thanks :)

**Comment by SaschaMann** [comment]
> could you please outline why below

Because horizontal scrolling sucks.

**Reply by iHiD** [reply]
An example will be needed for us to understand that :)

**Reply by SaschaMann** [reply]
![grafik](https://user-images.githubusercontent.com/20866761/89950119-3b9aec00-dc29-11ea-98fb-fdbee23825fb.png)

(Open in full screen to get a better idea of it)

Not the best example because it only hides comments behind the scrollbar, but I've seen it happen for code before too. I don't know what the width is atm, but it's like 50 chars too small at least.

**Reply by SaschaMann** [reply]
It would be fine to not have the code full-width if you made use of the entire screen but I'm guessing that has other downsides. I think ideally I'd want to be able to drag around panes/UI elements however I want like in VS Code.

**Reply by iHiD** [reply]
> It would be fine if you actually made use of the entire screen but I'm guessing that has other downsides

That is the full screen for most people using Exercism, so it needs to work at that resolution. Supporting other resolutions is a sort of secondary optimisation (and while I get that you might like that, it doesn't solve the core problem).

**Reply by iHiD** [reply]
> Because horizontal scrolling sucks

It strikes me that not being able to see the code I'm reviewing, the comment I'm responding to, and the response I've writing at the same time, is much more painful, than having some horizontal scrolling.

Maybe the reality here is that because you're on a large screen, this doesn't affect you so much?

**Reply by SaschaMann** [reply]
> It strikes me that **not being able to see the code I'm reviewing**, the comment I'm responding to, and the response I've writing at the same time, is much more painful, than having some horizontal scrolling.

But if there's horizontal scrolling, I **can't** see the code I'm reviewing, which is why it's a problem for me. 120ish is the standard width for many code bases nowadays, so I think it would still fit on people's screens while having enough space for UI elements. See e.g. https://www.theregister.com/2020/06/01/linux_5_7/

**Reply by SaschaMann** [reply]
> Maybe the reality here is that because you're on a large screen, this doesn't affect you so much?

I just tested it on my 2ndary screen (1280x1024) and the page doesn't make use of all the space available there either, despite being 700ish pixels smaller than a standard monitor.

![grafik](https://user-images.githubusercontent.com/20866761/89951199-42c2f980-dc2b-11ea-922e-37010691aa1a.png)

**Reply by kotp** [reply]
This is exactly why, the areas don't allow the full use of the width that may be available to the viewer.  This is true for many websites, preferring (for some strange reason) to have things in the middle of the screen, and making it a tad difficult to move things even in different windows left and right laid out.

**Reply by SaschaMann** [reply]
> This is true for many websites

It's better to limit the width of a line for reading text content for readability (see e.g. [Wikipedia](https://en.wikipedia.org/wiki/Line_length#Electronic_text)). I doubt that's an issue in this context, though, since it's largely not text content. 

> (for some strange reason)

More often than not the reason is that it allows them to fill the rest of the page with ads, like [this](https://i.redd.it/6b9axrqgr4z11.png).

**Reply by iHiD** [reply]
> the page doesn't make use of all the space available there either

It adds padding on the sides so that the content can actually breath and be readable/usable. You don't want things butting up against the scrollbar. Maybe the padding is too much, but it's also extremely standard.

> (for some strange reason)

The actual reason we do it is because it's much harder to make a design that works well at 3000px wide, 1080px wide and 400px wide (mobile). The amount of extra work for adding those larger breakpoints is hard. On pages of text, you don't want sentences that have 30-40 words on a line for example. So then each page ends up with it's own set of rules about how wide it should be, and things get very messy very quickly.

Take a look at the wikipedia page Sascha linked to for example - on my big screen it's unreadable because it's so long, whereas the [mobile version](https://en.m.wikipedia.org/wiki/Line_length#Electronic_text) looks great for me, because it has space either side. 

Compare:

<img width="1440" alt="Screenshot 2020-08-11 at 23 49 37" src="https://user-images.githubusercontent.com/286476/89956589-5cfdd700-dc2d-11ea-8b30-2303414ced83.png">
<img width="1440" alt="Screenshot 2020-08-11 at 23 49 45" src="https://user-images.githubusercontent.com/286476/89956596-5ff8c780-dc2d-11ea-9bf3-00012b7ac9f1.png">

I'm not arguing that one is right or wrong, and I appreciate that people can just choose to resize their browser to whatever width they like, but from decades of experience, it's much harder to make a website work well across a range from 400 - 3000px than it is to range from 400-1200 px.

**Reply by kotp** [reply]
So would it be reasonable to orient feedback at around 60 characters, and have the space reserved for a longer line for the code side?  At the moment, I believe it is about evenly divided. (Of course, then the feedback that has code would have the scrolling for the student, and the content is not downloadable for them.)

**Reply by iHiD** [reply]
> So would it be reasonable to orient feedback at around 60 characters, and have the space reserved for a longer line for the code side? 

Potentially yes. Although your subsequent comment in brackets is my concern.

Also, the views for the students and mentor don't need to be similar, as they serve different purposes and different mindsets.

The key need right now is to understand **why** the horizontal scrolling is a bigger problem than vertical scrolling. Why people feel it's better to be able to see none of the code while writing the comment (which is what I get in full-width mode) than seeing most of the code while writing the comment, but having to scroll if the code is too long. Why, for example, line-wrapping wouldn't be a better solution.

**Reply by SaschaMann** [reply]
> (which is what I get in full-width mode) 

Could you post an example of what it looks like for you, please?

> Why, for example, line-wrapping wouldn't be a better solution.

Completely personal preference, but I find line-wrapping works well for text, but makes code much harder to read because it typically breaks the indentation and splits up in the wrong places.

> Take a look at the wikipedia page Sascha linked to for example - on my big screen it's unreadable because it's so long, whereas the mobile version looks great for me, because it has space either side.

try putting another article next to it in mobile mode and you'll likely still have plenty of whitespace for aesthetics while making better use of the space :) E.g. one article with some definitions on the left, and another article that uses them.

**Reply by iHiD** [reply]
Sure. These screenshots are full size on my macbook. This solution is a good example, as it has long lines.

In full width mode, the code takes the full screen and I have to scroll down to see the new comment box. Once that first comment is posted, I wouldn't then see any code at the same time as future comments:

<img width="1414" alt="Screenshot 2020-08-12 at 01 00 31" src="https://user-images.githubusercontent.com/286476/89960620-6db34a80-dc37-11ea-9a75-9a2478108a73.png">
<img width="1241" alt="Screenshot 2020-08-12 at 01 00 37" src="https://user-images.githubusercontent.com/286476/89960626-7015a480-dc37-11ea-89bb-1100665b3a6d.png">

---

With LHS/RHS, I have to horizontally scroll to see a couple of the LOCs but I can write the comment while reading most of the code:

<img width="1174" alt="Screenshot 2020-08-12 at 01 00 57" src="https://user-images.githubusercontent.com/286476/89960665-8c194600-dc37-11ea-90e9-2905c185b66d.png">

**Reply by kotp** [reply]
> > So would it be reasonable to orient feedback at around 60 characters, and
> > have the space reserved for a longer line for the code side?
> 
> Potentially yes. Although your subsequent comment in brackets is my concern.

Yep, this area, what I would prefer is having independent vertical scroll, so I
can see comments when looking at parts of the code, so further on down when it
is side by side, and it is the 7th round of comments, I can't see the code where
I want to see it, perhaps line 4 or 10, while being way down on the right hand
side in the comments.  Independent vertical scrolling to have the context while
in the browser.

> Also, the views for the students and mentor don't need to be similar, as they
> serve different purposes and different mindsets.

I definitely see that being the case.

> The key need right now is to understand **why** the horizontal scrolling is a
> bigger problem than vertical scrolling. Why people feel it's better to be able
> to see none of the code while writing the comment (which is what I get in
> full-width mode) than seeing most of the code while writing the comment, but
> having to scroll if the code is too long. Why, for example, line-wrapping
> wouldn't be a better solution.

Line wrapping poses other problems if it is not apparent or obvious that it is a
line continuation.  I don't find that so much of a problem when it is indicated.
I really try to avoid reading code in the browser though, same with writing
comments, but we are stuck reading comments in the browser for the most part.

**Reply by taiyab** [reply]
I like some of the ideas coming out of this.

What it looks like is, the UI should maybe:

- Use full-width, edge-to-edge window
- Use width constraints/wrapping on non-code content
- Have the option to either split panes across or down (like a text editor)
- Have the option to adjust pane widths

It'll be difficult to maybe get all of this working perfectly, but in essence if the code never wraps, it makes things a lot easier (as the code pane can always take up the remainder of excess space).

**Reply by ynfle** [reply]
> > could you please outline why below
> 
> Because horizontal scrolling sucks.

I wonder if it differs for different computers and input devices?
I use a MacBook and the trackpad has excellent horizontal scrolling support

**Comment by wneumann** [comment]
I… didn't even know there was such a thing as full-width mode.

Having said that, I can't imagine I'd need it. Any code that I would need to keep scrolling side-to-side for, I'd probably download and look at in my text editor. For a quick glance, I don't really care if I need to scroll once or twice.

**Comment by tjarratt** [comment]
I too only discovered this feature today. It's interesting, and I'm trying it out, but I personally find that it is less helpful at the beginning of the tracks. Some of the later core problems and side problems benefit from this, where the code can frequently go past 80 characters.

Personally my workflow is to take the solution source code, copy it, and then use `pbpaste` to drop it into my own solution (which I have tracked by git) so that I can run the tests, try out any refactors before I propose them and make sure I'm not leading anyone astray.

Sometimes when I copy and paste I end up copying the first letter of the following word of column to the right (which is invariably an `M` from the phrase `Mentor discussion`. Having a button to automatically put the source code into my clipboard so I can bring it onto my local machine would be very helpful for my own workflow, but I don't know if anyone else does this.

**Reply by taiyab** [reply]
Interesting comments about side-by-side being more useful during lighter exercises.

Also, for sure, bringing a copy function into the code previewer is a great idea.

**Comment by Eiermann** [comment]
_For those people who have full-width enabled, could you please outline why below._

Since discovering full-width mode a few hundred mentorings ago I've never been tempted to go back. Losing context when looking at partial lines of code is significantly more frustrating for me than needing to scroll between code and comment. I would prefer wrapping to horizontal scrolling but I'm pretty content with the existing full-width mode.

**Comment by SleeplessByte** [comment]
I personally would want a ~125ch width box, with a line at 80 and 120ch (but maybe configurable _per language_).

This looks like this:

![Image showing some code in VS Code with a vertical divider at 80 and 120 characters](https://user-images.githubusercontent.com/1964376/90074888-6bff8a80-dcfc-11ea-828c-cb4d38502ef9.png)

Why? 

- It creates a reference point, allowing visual distancing from either side, which can be really helpful skimming code
- Long lines (over 120ch) are _terrible_ to read, or select using a mouse. Try selecting a bit of minimified code. Don't do it.
- For JavaScript and TypeScript I want to turn on prettier by default, which will wrap it at 80/120 anyway.

------

Why use full width mode?

![Non-full-width mode on half a screen](https://user-images.githubusercontent.com/1964376/90075351-70787300-dcfd-11ea-9164-1a9326a8ac10.png)

![Full-width mode on screen](https://user-images.githubusercontent.com/1964376/90075324-6191c080-dcfd-11ea-85e2-1b13b21b36e1.png)


Because half-screen mentoring is horrible when I'm _not_ using full width mode. 

-----

What is it really solving?

I want to be able to turn on and off the comment section more easily. I'd love it if comments could overlay, or maybe only the last one, or maybe each one in a box, or maybe all the previous ones and mine separately. Think: gmail email response which you can "pop out". This would solve most the issues for most people. 

And don't make me scroll for at least the first 80ch.

🤷‍♂️

**Reply by kotp** [reply]
Being able to easily switch between the two views is something that I would think would be nice.
